frame/support/procedural/src/pallet/expand/event.rs

```diff
@@ -41,18 +41,23 @@ pub fn expand_event(def: &mut Def) -> proc_macro2::TokenStream {
 	let frame_support = &def.frame_support;
 	let event_use_gen = &event.gen_kind.type_use_gen(event.attr_span);
 	let event_impl_gen= &event.gen_kind.type_impl_gen(event.attr_span);
+
 	let metadata = event.metadata.iter()
-		.map(|(ident, args, docs)| {
-			let name = format!("{}", ident);
-			quote::quote_spanned!(event.attr_span =>
-				#frame_support::event::EventMetadata {
-					name: #frame_support::event::DecodeDifferent::Encode(#name),
-					arguments: #frame_support::event::DecodeDifferent::Encode(&[
-						#( #args, )*
-					]),
-					documentation: #frame_support::event::DecodeDifferent::Encode(&[
-						#( #docs, )*
-					]),
+		.map(|event| {
+			let name = format!("{}", event.name);
+			let args = event.args
+				.iter()
+				.map(|(ty, name)| {
+					quote::quote!(
+						#frame_support::metadata::v13::TypeSpec::new::<#ty>(#name)
```

Comment:
here `name` is the type tokens convert to string: `clean_type_string(&ty.to_token_stream().to_string())`. I'm not sure it is really meaningfull

If the event is declare using name field:
```
pub enum Event {
A { a: u32 },
```
Then `a` as a name is valuable but this is not what is done here.

Also instead of giving the event type info by giving only the argument of each variant, why not giving the type info in a whole. And separately giving the doc for each variant ?
Maybe your current method is better for UI, I just wonder what is best, like we can do `meta_type::<Event>()`.

Same question can be asked for `Call` enum instead of giving the type of each argument of each call. We could give the type of the Call enum as a whole.

Reply:
> Maybe your current method is better for UI, I just wonder what is best, like we can do meta_type::<Event>().

At first I thought it was better to have the new structure be as similar to the old as possible, just replacing type strings with `scale-info` types. 

However I now agree it is probably better just to utilize the type derives on the call, event and error enums. I've added doc capturing to `scale-info` https://github.com/paritytech/scale-info/pull/87 so now should have everything we need. 

Then we can simply replace all the extra metadata types for calls, events and errors with type specs: https://github.com/paritytech/frame-metadata/pull/11.

Let me know what you think.

Reply:
I personally like it, (except maybe for calls as mentionned in the linked PR), but maybe some other people can have different opinion, and see other usecase and extension of metadata which would require going back to the old structure.

Reply:
I think it's too new for anyone to have strong opinions about what serves future use cases best. What @thiolliere suggests seems rather elegant though so let's go with that imo.

Reply:
I've now integrated the changes from https://github.com/paritytech/frame-metadata/pull/11 into this PR. Main points:

- Replaces error and event custom metadata types with `scale-info` enum meta types
- Points to `Call` enum  `scale-info` metatype but maintains existing call metadata types for extra info.
- Removes `OuterEventMetadata` (it was unused here). Now that I mention it, I could add a reference to the outer event and call types from the top level runtime metadata. I'll add that to https://github.com/paritytech/frame-metadata/pull/11.
- Also renames some types to pallet instead of module.

Please review that PR and let me know what you think of the refactor.